### DISCUSSION TITLE
get_textpage_ocr() No OCR support: TESSDATA_PREFIX not set

### DISCUSSION THREAD

**Discussion by yonglee7015** [pymupdf/PyMuPDF#2439] [Looking for help] [answered]
get_textpage_ocr() not work
Google Colab.  I  follow instruction as below .

> Enabling Integrated OCR Support
> If you do not intend to use this feature, skip this step. Otherwise, it is required for both installation paths: from wheels and from sources.
> 
> PyMuPDF will already contain all the logic to support OCR functions. But it additionally does need Tesseract’s language support data, so installation of Tesseract-OCR is still required.
> 
> The language support folder location must currently [[1]](https://pymupdf.readthedocs.io/en/latest/installation.html#f1) be communicated via storing it in the environment variable "TESSDATA_PREFIX".
> 
> So for a working OCR functionality, make sure to complete this checklist:
> 
> Install Tesseract.
> 
> Locate Tesseract’s language support folder. Typically you will find it here:
> Windows: C:Program FilesTesseract-OCRtessdata
> 
> Unix systems: /usr/share/tesseract-ocr/4.00/tessdata
> 
> Set the environment variable [TESSDATA_PREFIX](https://pymupdf.readthedocs.io/en/latest/functions.html#TESSDATA_PREFIX)
> Windows: set TESSDATA_PREFIX=C:Program FilesTesseract-OCRtessdata
> 
> Unix systems: export TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata

!apt install tesseract-ocr
 set environment  TESSDATA_PREFIX='/usr/share/tesseract-ocr/4.00/tessdata' 

RuntimeError: No OCR support: TESSDATA_PREFIX not set

**Comment by JorjMcKie** [comment]
Shouldn't you use the command `!export TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata` in this environment?

**Reply by yonglee7015** [reply]
Then what should I do in Google Colab? 
 I use !export TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata
But it still show this runtime error

![image](https://github.com/pymupdf/PyMuPDF/assets/59268746/9aa79cae-2fdf-4041-a02a-fd5605d0790a)

**Reply by JorjMcKie** [reply]
Hm, I looked around in the Colab documentation. It seems that the Python environment is already running when you enter Colab. And that environment variable has to be set before Python is running.
So an idea could be to start a subprocess to execute the `export` command and then also modify `os.environ` and set the variable ... IAW something like this:
```
import subprocess,os
subprocess.call(["export", "TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata"])
os.environ["TESSDATA_PREFIX"] = "/usr/share/tesseract-ocr/4.00/tessdata"
```
Maybe, considering the `!export` may be doing exactly the same, calling subprocess is not even necessary ...

**Reply by JorjMcKie** [reply, marked as answer]
Ah, and in PyMuPDF there is a variable being set in `__init__.py` that checks the `os.environ` variable, so the above must happen before import fitz.

**Reply by JorjMcKie** [reply]
Thinking more about it, you should be fine simply setting `fitz.TESSDATA_PREFIX = "/usr/share/tesseract-ocr/4.00/tessdata"`

**Reply by yonglee7015** [reply]
> Hm, I looked around in the Colab documentation. It seems that the Python environment is already running when you enter Colab. And that environment variable has to be set before Python is running. So an idea could be to start a subprocess to execute the `export` command and then also modify `os.environ` and set the variable ... IAW something like this:
> 
> ```
> import subprocess,os
> subprocess.call(["export", "TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata"])
> os.environ["TESSDATA_PREFIX"] = "/usr/share/tesseract-ocr/4.00/tessdata"
> ```
> 
> Maybe, considering the `!export` may be doing exactly the same, calling subprocess is not even necessary ...

[Errno 2] No such file or directory: 'export'
subprocess  not work

**Reply by yonglee7015** [reply]
> Thinking more about it, you should be fine simply setting `fitz.TESSDATA_PREFIX = "/usr/share/tesseract-ocr/4.00/tessdata"`

still not work. 

RuntimeError: No OCR support: TESSDATA_PREFIX not set

**Reply by yonglee7015** [reply]
> Ah, and in PyMuPDF there is a variable being set in `__init__.py` that checks the `os.environ` variable, so the above must happen before import fitz.

**yes ,must happen before import fitz.**

import os
os.environ["TESSDATA_PREFIX"] = "/usr/share/tesseract-ocr/4.00/tessdata"


import fitz

Then it works. thank you.  Runtime error disappeared.

but the OCR result is so terrible. 
set full = True, it seems not so bad

**Reply by yonglee7015** [reply]
Hi @JorjMcKie, why are there so many useless characters?Maybe it comes from the picture of garment.  Is there a better way to eliminate these characters? 

MINUMUM CONSTRUCTION STANDARDS SWEATERS ees ‘ 
Vendor is responsible for meeting minimum construction standards found on our Vendor Extranet j —— \ CONSTRUCTION DETAILS / TRIMS Me ¥ ; ’ *ShoulderSeams: 1/4” DTM stretchtape +f NG ‘ LL Le fe applied with4 thread averlock. t7, \t ASAEA *Hanger Loops: 1/4” wideDTM double sided satin ~ po \ Nee F 7 ribbon, 8” longon fold. Place at 1/2” below armhole b 1 Ny , aAAA J 
latsideseams. Ends must be clean finished. “G/ N) | ee ea hy od “Main Label: Tack at upper comers with minimum 2 | iA +) f A [A tA i 3 passes with 2 ends of poly corethread tex 40. f v/ | Xe tek 7, L) 4 Collar/Front placket-: Self facing Ang, | . N efhah A Cuff: Self start = Ms NAT AS /S 7S” Sweep: Self start wi Pree yi ; * Belt: Birdseye jacquard backed See 
withDTMgrosgrain. q ; 

![image](https://github.com/pymupdf/PyMuPDF/assets/59268746/c9eaed80-5246-43b5-b4f7-5ec07a760f9b)

**Reply by JorjMcKie** [reply]
First of all, congratulations on getting OCR working in Colab!
Otherwise: yes you started with an unfortunately bad example. You can still play a bit with the resolution / DPI.
